lens/private/list/flatten.rkt

```diff
@@ -0,0 +1,204 @@
+#lang racket/base
+
+(provide append*-lens flatten/depth-lens flatten/depth unflatten/depth)
+
+(require fancy-app lens racket/list racket/match)
+
+(module+ test
+  (require rackunit syntax/parse lens/private/test-util/test-lens))
+
+;; restore-structure : (Listof (Listof A)) (Listof B) -> (Listof (Listof B))
+;; Takes a list of lists and a list and un-flattens the flattened
+;; argument according to the structure of the structure arguement.
+;; The length of the flattened list must be the same as the length
+;; of (append* structure).
+(define (restore-structure structure flattened)
+  (match structure
+    [(list)
+     (unless (empty? flattened)
+       (error 'append*-lens "flattened list is too long to match the structure"))
+     structure]
+    [(cons s-lst s-rst)
+     (define-values [f-lst f-rst]
+       (split-at flattened (length s-lst)))
+     (cons f-lst (restore-structure s-rst f-rst))]))
+
+;; append*-lens : (Lens (Listof (Listof Any)) (Listof Any))
+;; where the only valid views are lists with the same length as the
+;; result of applying append* to the target.
+;; This is conceptually equivalent to (flatten/depth-lens 2)
+(define append*-lens
+  (make-lens
+   append*
+   restore-structure))
+
+;; (define-type (Listof* A n)
+;;   (cond [(zero? n) A]
+;;         [else (Listof* (Listof A) (sub1 n))]))
+
+;; flatten/depth-lens : (Lens (Listof* Any n) (Listof Any))
+;; where the only valid views are lists with the same length as the
+;; result of (flatten/depth n target)
+(define (flatten/depth-lens n)
+  (make-lens
+   (flatten/depth n _)
+   (unflatten/depth n _ _)))
+
+;; flatten/depth : n (Listof* A n) -> (Listof A)
+(define (flatten/depth n lst*)
+  (cond [(zero? n) (list lst*)]
```

Comment:
I was using `cond` because I'm thinking of natural numbers as an enumeration, one of zero or a successor of another number. The design recipe uses `cond` for enumerations for two cases as well as for more. 

My style for natural-number-processing functions is somewhat based on [this section of HtDP/2e](http://www.ccs.neu.edu/home/matthias/HtDP2e/part_two.html#%28part._sec~3anats%29).